### PR TITLE
Preload the projects into the memory before starting the server

### PR DESCRIPTION
**Pull request by RitwikGopi**
**Proposed changes**: Preload the projects into the memory before starting the server as per user requirement through command line argument. By default, no projects will be loaded. If specified an argument `--pre_load` as `all`, all the projects will be preloaded with the latest model. Otherwise, the user can specify a list of projects and those alone will be preloaded.

**Status (please check what you already did)**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog

### CODE REVIEW

File: rasa_nlu/data_router.py [hunk 1]
```diff
@@ -170,6 +169,11 @@ def _create_project_store(self, project_dir):
                     remote_storage=self.remote_storage)
         return project_store
 
+    def _pre_load(self, projects):
+        logger.debug("loading %s", projects)
+        filtered = [p for p in self.project_store if p in projects]
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: mhm I think we can combine the filter and map as we don't need their results:
```python
for p in self.project_store:
  if p in projects:
    self.project_store[p].load_model()
```

File: rasa_nlu/project.py [hunk 2]
```diff
@@ -126,6 +126,25 @@ def parse(self, text, time=None, requested_model_name=None):
 
         return response, model_name
 
+    def load_model(self):
+        self._begin_read()
+        status = False
+        model_name = self._dynamic_load_model()
+        logger.debug('Loaidng model %s', model_name)
```
Comment (tmbo) [inline rasa_nlu/project.py]: 'Loading'

File: rasa_nlu/server.py [hunk 3]
```diff
@@ -31,12 +31,20 @@ def create_argument_parser():
...
                         choices=['wit', 'luis', 'dialogflow'],
-                        help='which service to emulate (default: None i.e. use '
-                             'simple built in format)')
+                        help='which service to emulate (default: None i.e. use'
+                             ' simple built in format)')
     parser.add_argument('-P', '--port',
                         type=int,
                         default=5000,
                         help='port on which to run server')
+    parser.add_argument('-p', '--pre_load',
```
Comment (tmbo) [inline rasa_nlu/server.py]: let's not do a short argument (`-p`) and just use `--pre_load` (as `-p` might easily be confused with the port one)

**Comment by RitwikGopi** [comment]
@tmbo thanks for the suggestions. I will make the changes.

**Comment by RitwikGopi** [comment]
@tmbo all the suggested changes have been done. Regarding doc. The Server Parameters in the doc are generated using `program-output`. Should I add an additional section dedicated to preload alone in the doc?

**Comment by tmbo** [comment]
You are right, the description in the help text for the parameter you created is very good. So let's keep it like that for now and if it comes up in the community we can still add a paragraph dedicated to it. 

Thanks a lot for all the effort you put into getting this merged :rocket:

**Comment by parthsharma1996** [comment]
@RitwikGopi I'm unable to use this as I described in #1193 . Could you help me out with finding the problem? Can you also give an example usage?

**Comment by antoinesauray** [comment]
Also having the issue. My models are stored on S3. I get this error when using `--pre_load all`

`botocore.exceptions.ClientError: An error occurred (404) when calling the HeadObject operation: Not Found`

**Pull request merged**